core/drivers/crypto/marvell/authenc.c

```diff
@@ -0,0 +1,490 @@
+		memcpy(encdata_buf + iv + ctx->aad_cur_idx + dfinal->src.length,
+		       dfinal->tag.data, dfinal->tag.length);
+
+		ret = mrvl_ehsm_aes_gcm_final(encdata_buf, dfinal->src.length,
+					      decdata_buf, dfinal->dst.length,
+					      ctx->is_new);
+
+		if (ret == TEE_SUCCESS)
+			memcpy(dfinal->dst.data, decdata_buf + iv +
+			       ctx->aad_cur_idx, dfinal->dst.length);
+
+	} else {
+		buf_len = dfinal->src.length + dfinal->tag.length;
+
+		encdata_buf = mem_alloc(buf_len);
+		if (!encdata_buf) {
+			ret = TEE_ERROR_OUT_OF_MEMORY;
+			goto out;
+		}
+
+		decdata_buf = mem_alloc(buf_len);
+		if (!decdata_buf) {
+			ret = TEE_ERROR_OUT_OF_MEMORY;
+			goto out_dec_buf_err;
+		}
+
+		memcpy(encdata_buf, dfinal->src.data, dfinal->src.length);
+		memcpy(encdata_buf + dfinal->src.length, dfinal->tag.data,
+		       dfinal->tag.length);
+
+		ret = mrvl_ehsm_aes_gcm_final(encdata_buf, dfinal->src.length,
+					      decdata_buf, dfinal->dst.length,
+					      ctx->is_new);
+
+		if (ret == TEE_SUCCESS)
+			memcpy(dfinal->dst.data, decdata_buf,
+			       dfinal->dst.length);
+	}
+
+	free(decdata_buf);
+out_dec_buf_err:
+	free(encdata_buf);
+out:
+	mrvl_ehsm_cryp_unlock();
+	return ret;
+}
+
+static void mrvl_ae_final(void *ctx __unused)
+{
+}
+
+static void mrvl_ae_free(void *ctx)
+{
+	struct mrvl_ae_ctx *c = to_mrvl_ae_ctx(ctx);
+	struct mrvl_cryp_context *cryp_ctx = &c->cryp_ctx;
+
+	mrvl_ehsm_cryp_lock();
+	mrvl_ehsm_aes_context_release(cryp_ctx->context_id);
+	mrvl_ehsm_cryp_unlock();
+
+	mrvl_ehsm_aes_cryp_release();
+
+	free(cryp_ctx->iv_data);
+	free(cryp_ctx->aad_data);
+	free(cryp_ctx->key);
+	free(c);
+}
+
+static void mrvl_ae_copy_state(void *dst_ctx, void *src_ctx)
+{
+	struct mrvl_ae_ctx *src = to_mrvl_ae_ctx(src_ctx);
+	struct mrvl_ae_ctx *dst = to_mrvl_ae_ctx(dst_ctx);
+
+	memcpy(dst, src, sizeof(*dst));
+}
+
+/*
+ * Allocate the SW authenc data context
+ *
+ * @ctx   [out] Caller context variable
+ * @algo  Algorithm ID of the context
+ */
+static TEE_Result mrvl_ae_allocate(void **ctx, uint32_t algo)
+{
+	struct mrvl_ae_ctx *c = NULL;
+
+	if (algo != TEE_ALG_AES_GCM)
+		return TEE_ERROR_NOT_IMPLEMENTED;
+
+	if (!mrvl_ehsm_aes_cryp_allowed()) {
+		DMSG("%s ehsm aes busy, algo: %x\n", __func__, algo);
+		return TEE_ERROR_NOT_IMPLEMENTED;
+	}
+
+	c = calloc(1, sizeof(*c));
+	if (!c) {
+		mrvl_ehsm_aes_cryp_release();
+		return TEE_ERROR_OUT_OF_MEMORY;
+	}
+
+	*ctx = &c->a_ctx;
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * Registration of the Authenc Driver
+ */
+static struct drvcrypt_authenc mrvl_gcm = {
+	.alloc_ctx = &mrvl_ae_allocate,
+	.update_aad = &mrvl_ae_update_aad,
+	.update_payload =  &mrvl_ae_update_payload,
+	.enc_final = &mrvl_ae_enc_final,
+	.dec_final = &mrvl_ae_dec_final,
+	.free_ctx = &mrvl_ae_free,
+	.init = &mrvl_ae_initialize,
+	.final = &mrvl_ae_final,
+	.copy_state = &mrvl_ae_copy_state,
+};
+
+static TEE_Result mrvl_register_authenc(void)
+{
+	return drvcrypt_register_authenc(&mrvl_gcm);
```

Comment:
Do you want to panic if `drvcrypt_register_authenc()` fails? Otherwise, upon such failure, OP-TEE would always fallback to a software crypto implementation.